src/qtwebengine/dataprovider.cpp

```diff
@@ -26,13 +72,23 @@ void DataProvider_QWebEngine::requestStarted( QWebEngineUrlRequestJob *request )
         return;
     }
 
-    QString mimetype;
+    QByteArray mimetype;
 
-    // Specify the encoding in case the page content is not UTF-8
+    // We must specify the proper MIME type for the page to display correctly.
+    // The HTML and XML files correspond to "text/html";
+    // for other types "application/octet-stream" is sufficient.
+    // In addition, for "text/html", a "meta" tag is added specifying the text encoding.
+    // This is the easiest and most stable way to set the encoding.
     if ( htmlfile )
-        mimetype = QString( "text/html; charset=%1" ) .arg( ::mainWindow->chmFile()->currentEncoding() );
+    {
+        mimetype = "text/html";
+        buf.prepend(QString( "<META http-equiv='Content-Type' content='text/html; charset=%1'>" )
+                    .arg( ::mainWindow->chmFile()->currentEncoding() ).toLatin1() );
```

Comment:
Note that your prepend will go before `<html>` tag, which may internally have another http-equiv.